nix/flake.nix

```diff
@@ -1,33 +1,143 @@
+      flake = false;
+    };
   };
 
-  outputs = { self, nixpkgs, flake-utils }:
-    let kmonad = import ./default.nix;
-    in flake-utils.lib.eachDefaultSystem (system:
-      let pkgs = nixpkgs.legacyPackages.${system};
-      in rec {
-        packages.kmonad = pkgs.haskellPackages.callPackage kmonad { };
+  outputs = { self, nixpkgs, Karabiner-DriverKit-VirtualHIDDevice }:
+    let
+      # List of supported systems:
+      supportedSystems = [ "x86_64-linux" "x86_64-darwin" ];
 
-        defaultPackage = packages.kmonad;
+      # List of supported compilers:
+      supportedCompilers = [
+        "ghc8107"
+        "ghc902"
+        "ghc922"
+      ];
 
-        devShell = pkgs.haskellPackages.shellFor {
-          packages = _: [ packages.kmonad ];
-          withHoogle = true;
-          buildInputs = [ pkgs.haskellPackages.cabal-install ];
-        };
-      }) // {
-        nixosModule = ({ ... }: {
-          nixpkgs.overlays = [ self.overlay ];
-          imports = [ ./module-base.nix ];
-        });
-
-        overlay = final: prev: {
-          kmonad = final.haskellPackages.callPackage kmonad { };
+      # Function to generate a set based on supported systems:
+      forAllSystems = f:
+        nixpkgs.lib.genAttrs supportedSystems (system: f system);
+
+      # Attribute set of nixpkgs for each system:
+      nixpkgsFor = forAllSystems (system: import nixpkgs { inherit system; });
+
+      # A source file list cleaner for Haskell programs:
+      haskellSourceFilter = src:
+        nixpkgs.lib.cleanSourceWith {
+          inherit src;
+          filter = name: type:
+            let baseName = baseNameOf (toString name); in
+            nixpkgs.lib.cleanSourceFilter name type &&
+            !(
+              baseName == "dist-newstyle"
+              || nixpkgs.lib.hasPrefix "." baseName
+            );
         };
+
+      # A fake version of git which can be used during the build to
+      # embed the current revision into the binary despite the .git
+      # directory not being available:
+      fakeGit = pkgs: pkgs.writeShellScriptBin "git" ''
+        echo ${self.rev or "dirty"}
+      '';
+
+      # The package derivation:
+      derivation = pkgs: haskell: (
+        # TODO: make this os-conditional and remove the --no-check flag in the future when tests are fixed
+        haskell.callCabal2nixWithOptions "kmonad" (haskellSourceFilter ../.) "--flag=dext --no-check" { }
+      ).overrideAttrs (orig: {
+        buildInputs = orig.buildInputs ++ [ (fakeGit pkgs) ] ++
+          (pkgs.lib.optional pkgs.stdenv.hostPlatform.isDarwin pkgs.darwin.IOKit);
+      } // (pkgs.lib.optionalAttrs pkgs.stdenv.hostPlatform.isDarwin {
+        configureFlags = orig.configureFlags ++ [
+          "--extra-include-dirs=c_src/mac/Karabiner-DriverKit-VirtualHIDDevice/src/Client/vendor/include"
+          "--extra-include-dirs=c_src/mac/Karabiner-DriverKit-VirtualHIDDevice/include/pqrs/karabiner/driverkit"
```

Comment:
I tried removing those flags but the build fails without them. It's probably because of the environment Nix builds things under, but I'm not going to worry about it since including them doesn't seem to cause any issues.

Reply:
In `kmonad.cabal` file, there is a check for the `kext`/`dext` flag that will add `extra-include-dirs` with the correct path. So if it doesn't work, then it is a bug.

ref: https://github.com/kmonad/kmonad/blob/1d2276069fd781c4518082f1b36f3b2880667e79/kmonad.cabal#L181-L195

Reply:
:facepalm: I'm dumb, it was failing because of the issue I documented in 5162d542b0f9b2110cd8b760717c5699edbbea18... Thanks! Fixed in the latest force-push.